### PR TITLE
Added ability to handle empty model in timepicker

### LINKED ISSUE #1114: TimePicker shows current time when model is empty.
TimePicker shows current time when the model is null or cleared. On the contrary, DatePicker is blank when model is empty, which is the correct approach. It is impossible to tell when the TimePicker model is not set by looking at the control. TimePicker should be blank when model is not set.

### PR DESCRIPTION
**Pull request by fmccown**
Model can be empty so user must type values into the hour and minute
text boxes. This closes #1114.

### CODE REVIEW

File: src/timepicker/timepicker.js [hunk 1]
```diff
@@ -143,6 +143,9 @@ angular.module('ui.bootstrap.timepicker', [])
   }
 
   function pad(value) {
+    if (value === null) {
```
Comment (wesleycho) [inline src/timepicker/timepicker.js]: `if (!value) {` may be a better check here.

File: src/timepicker/timepicker.js [hunk 2]
```diff
@@ -231,9 +237,12 @@ angular.module('ui.bootstrap.timepicker', [])
     };
 
     hoursInputEl.bind('blur', function(e) {
-      if (!$scope.invalidHours && $scope.hours < 10) {
+      ngModelCtrl.$setTouched();
+      if ($scope.hours === null || $scope.hours === "") {
```
Comment (wesleycho) [inline src/timepicker/timepicker.js]: Change to `if (!$scope.hours) {`

File: src/timepicker/timepicker.js [hunk 3]
```diff
@@ -255,7 +267,10 @@ angular.module('ui.bootstrap.timepicker', [])
     };
 
     minutesInputEl.bind('blur', function(e) {
-      if (!$scope.invalidMinutes && $scope.minutes < 10) {
+      ngModelCtrl.$setTouched();
+      if ($scope.minutes === null) {
```
Comment (wesleycho) [inline src/timepicker/timepicker.js]: Change to `if(!$scope.minutes) {`

File: src/timepicker/timepicker.js [hunk 4]
```diff
@@ -300,17 +315,23 @@ angular.module('ui.bootstrap.timepicker', [])
...
   function updateTemplate(keyboardChange) {
-    var hours = selected.getHours(), minutes = selected.getMinutes();
-
-    if ($scope.showMeridian) {
-      hours = (hours === 0 || hours === 12) ? 12 : hours % 12; // Convert 24 to 12 hour system
-    }
+    if (ngModelCtrl.$modelValue == null) {
+      $scope.hours = null;
+      $scope.minutes = null;
+	  $scope.meridian = meridians[0];
```
Comment (wesleycho) [inline src/timepicker/timepicker.js]: Indentation is off here

Comment (wesleycho) [inline src/timepicker/timepicker.js]: Actually, seems like the whole block's indentation is off.

File: src/timepicker/timepicker.js [hunk 5]
```diff
@@ -353,8 +374,15 @@ angular.module('ui.bootstrap.timepicker', [])
...
   $scope.toggleMeridian = function() {
+    var minutes = getMinutesFromTemplate(),
+        hours = getHoursFromTemplate();
+    
     if (!$scope.noToggleMeridian()) {
-      addMinutesToSelected(12 * 60 * (selected.getHours() < 12 ? 1 : -1));
+      if (angular.isDefined(minutes) && angular.isDefined(hours)) {
+        addMinutesToSelected(12 * 60 * (selected.getHours() < 12 ? 1 : -1));
+      } else {
+        $scope.meridian = $scope.meridian == meridians[0] ? meridians[1] : meridians[0];
```
Comment (wesleycho) [inline src/timepicker/timepicker.js]: use `===` here.

File: template/timepicker/timepicker.html [hunk 6]
```diff
@@ -8,11 +8,11 @@
     </tr>
     <tr>
       <td class="form-group" ng-class="{'has-error': invalidHours}">
-        <input style="width:50px;" type="text" ng-model="hours" ng-change="updateHours()" class="form-control text-center" ng-readonly="::readonlyInput" maxlength="2">
+        <input style="width:50px;" type="text" placeholder="HH" ng-model="hours" ng-change="updateHours()" class="form-control text-center" ng-readonly="::readonlyInput" maxlength="2">
```
Comment (wesleycho) [inline template/timepicker/timepicker.html]: Remove the `placeholder` change

File: template/timepicker/timepicker.html [hunk 7]
```diff
@@ -8,11 +8,11 @@
...
       </td>
       <td>:</td>
       <td class="form-group" ng-class="{'has-error': invalidMinutes}">
-        <input style="width:50px;" type="text" ng-model="minutes" ng-change="updateMinutes()" class="form-control text-center" ng-readonly="::readonlyInput" maxlength="2">
+        <input style="width:50px;" type="text" placeholder="MM" ng-model="minutes" ng-change="updateMinutes()" class="form-control text-center" ng-readonly="::readonlyInput" maxlength="2">
```
Comment (wesleycho) [inline template/timepicker/timepicker.html]: Ditto

**Comment by anthonmanix** [comment]
I don't think he should change value === null to !value, because that won't work for 0 hours and 0 minutes.

**Comment by wesleycho** [comment]
Good point - however, the other aspects are still issues.

**Comment by wesleycho** [comment]
Even with code style cleaned up, this PR fails tests.

It needs to be worked on before this can be accepted.

**Comment by ftorghele** [comment]
the PR is green now

**Comment by wesleycho** [comment]
History needs to be cleaned up - there should not be any merge commits.

**Pull request closed without merging**